### ISSUE TITLE
Custom dataset training

### ISSUE THREAD

**Issue by sheilabam** [chengtan9907/OpenSTL#31]
It is a great honor to learn your code, and the results obtained by reproducing the moving-mnist dataset are highly consistent with the results in your literature. The work I am currently doing is to imitate the movement and prediction of mice predation, and create a custom data set with a shape and data structure similar to the moving-mnist test set mnist_test_seq.npy format. The data loader code is as follows, also I give an example of my dataset, but it is confusing surprisingly, after several parameter adjustments, the training loss and test loss both decreased from thousands, and the image quality evaluation index was very large (7-digit MSE). It is indeed difficult to solve the trouble, sincerely request I would be very grateful for your help.

import cv2
import gzip
import numpy as np
import os
import random

import torch
import torch.nn.functional as F
import torchvision
import torchvision.transforms as transforms
from torch.utils.data import Dataset, DataLoader
from PIL import Image

from openstl.datasets.utils import create_loader

def load_mnist(root, data_name='mnist'):
    # Load train dataset.
    file_map = {
        'mnist': 'moving_mnist/mnist_train_seq.npy',
    }
    path = os.path.join(root, file_map[data_name])
    mnist = np.load(path)
    mnist = mnist.transpose(1, 0, 2, 3)
    return mnist

def load_fixed_set(root, data_name='mnist'):
    # Load the test dataset
    file_map = {
        'mnist': 'moving_mnist/mnist_test_seq.npy',
    }
    path = os.path.join(root, file_map[data_name])
    dataset = np.load(path)
    dataset = dataset.transpose(1, 0, 2, 3)
    
    return dataset

class MovingMNIST(Dataset):

    def __init__(self, root, is_train=True, data_name='mnist',
                 n_frames_input=10, n_frames_output=10,
                 transform=None, target_transform=None, use_augment=False):
        super(MovingMNIST, self).__init__()

        self.dataset = None
        self.root = os.path.expanduser(root)
        self.transform = transform
        self.target_transform = target_transform
        self.is_train = is_train
        self.data_name = data_name
        
        if self.is_train:
            self.mnist = load_mnist(root, data_name)
        else:
            self.dataset = load_fixed_set(root, data_name)

        self.n_frames_input = n_frames_input
        self.n_frames_output = n_frames_output
        self.n_frames_total = self.n_frames_input + self.n_frames_output
        self.use_augment = use_augment

        self.mean = 0
        self.std = 1

    def _augment_seq(self, imgs, crop_scale=0.94):
        """Augmentations for video"""
        _, _, h, w = imgs.shape  # original shape, e.g., [10, 1, 64, 64]
        imgs = F.interpolate(imgs, scale_factor=1 / crop_scale, mode='bilinear')
        _, _, ih, iw = imgs.shape
        # Random Crop
        x = np.random.randint(0, ih - h + 1)
        y = np.random.randint(0, iw - w + 1)
        imgs = imgs[:, :, x:x + h, y:y + w]
        # Random Flip
        if random.randint(-2, 1) > 0:
            imgs = torch.flip(imgs, dims=(2, 3))  # rotation 180
        elif random.randint(-2, 1) > 0:
            imgs = torch.flip(imgs, dims=(2,))  # vertical flip
        elif random.randint(-2, 1) > 0:
            imgs = torch.flip(imgs, dims=(3,))  # horizontal flip
        return imgs

    def __getitem__(self, index):
        # need to iterate over time
        def _transform_time(data):
            new_data = None
            for i in range(data.size(0)):
                img = Image.fromarray(data[i].numpy(), mode='L')
                new_data = self.transform(img) if new_data is None else torch.cat([self.transform(img), new_data], dim=0)
            return new_data

        if self.is_train:
            input, output = self.mnist[index, :self.n_frames_input], self.mnist[index, self.n_frames_input:self.n_frames_total]
        else:
            input, output = self.dataset[index, :self.n_frames_input], self.dataset[index, self.n_frames_input:self.n_frames_total]

        if self.transform is not None:
            input = _transform_time(input)
        if self.target_transform is not None:
            output = _transform_time(output)
            
        # Reshape the input and output tensors
        input = torch.from_numpy(np.expand_dims(input, axis=1)).float()
        output = torch.from_numpy(np.expand_dims(output,  axis=1)).float()

        return input, output

    def __len__(self):
        if self.is_train:
            return len(self.mnist)
        else:
            return len(self.dataset)

def load_data(batch_size, val_batch_size, data_root, num_workers=4, data_name='mnist',
              pre_seq_length=10, aft_seq_length=10, in_shape=[10, 1, 64, 64],
              distributed=False, use_augment=False, use_prefetcher=False, drop_last=True):
    image_size = in_shape[-1] if in_shape is not None else 64
    train_set = MovingMNIST(root=data_root, is_train=True, data_name=data_name,
                            n_frames_input=pre_seq_length,
                            n_frames_output=aft_seq_length, use_augment=False)
    test_set = MovingMNIST(root=data_root, is_train=False, data_name=data_name,
                           n_frames_input=pre_seq_length,
                           n_frames_output=aft_seq_length, use_augment=False)

    dataloader_train = create_loader(train_set,
                                     batch_size=batch_size,
                                     shuffle=True, is_training=True,
                                     pin_memory=True, drop_last=True,
                                     distributed=distributed, use_prefetcher=use_prefetcher)
    dataloader_vali = create_loader(test_set,
                                    batch_size=val_batch_size,
                                    shuffle=False, is_training=False,
                                    pin_memory=True, drop_last=drop_last,
                                    distributed=distributed, use_prefetcher=use_prefetcher)
    dataloader_test = create_loader(test_set,
                                    batch_size=val_batch_size,
                                    shuffle=False, is_training=False,
                                    pin_memory=True, drop_last=drop_last,
                                    distributed=distributed, use_prefetcher=use_prefetcher)

    return dataloader_train, dataloader_vali, dataloader_test

if __name__ == '__main__':
    from openstl.utils import init_dist
    os.environ['LOCAL_RANK'] = str(0)
    os.environ['RANK'] = str(0)
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12357'
    dist_params = dict(launcher='pytorch', backend='nccl', init_method='env://', world_size=1)
    init_dist(**dist_params)

    dataloader_train, _, dataloader_test = \
        load_data(batch_size=4,
                  val_batch_size=4,
                  data_root='../../data/',
                  num_workers=4,
                  data_name='mnist',
                  pre_seq_length=10,
                  aft_seq_length=10,
                  distributed=False,
                  use_prefetcher=False)

    print(len(dataloader_train), len(dataloader_test))
    for item in dataloader_train:
        print(item[0].shape, item[1].shape)
        break
    for item in dataloader_test:
        print(item[0].shape, item[1].shape)
        break
![捕获](https://github.com/chengtan9907/OpenSTL/assets/127728617/36574d45-0863-4483-b490-416cc15ef1d8)

**Comment by chengtan9907** [comment]
I believe if you can visualize correctly, it shouldn't be a problem with data dimensions. It seems more like an issue with the data scale. In the `MovingMNIST` class of the Moving MNIST dataset, we defined the mean and std as 0 and 1 because the pixel values of Moving MNIST are in the range [0, 1], and we don't need to rescale them when calculating metrics. I'm not sure about the scale of your data.

And what do you think, @Lupin1998 ?

**Comment by sheilabam** [comment]
Thank you for your guidance. My current dataset includes 5212 samples (after 10 kinds of data amplification such as horizontal flip, vertical flip, and angle rotation), and the data scale is really not enough.

**Comment by chengtan9907** [comment]
I'd also like to ask, is your training loss big? It seems that you have trained a correct model, so the issue might be related to the metric evaluation. You can check if there is any rescaling issue during the metric evaluation phase.

**Comment by sheilabam** [comment]
Hello, my training loss is very large. The loss of training set and test set start to drop from about 3200. The training set can drop to several hundred after 200 rounds, the test set loss starts to decline, after about 30 rounds, it starts to rise and remains at 2000.

**Comment by chengtan9907** [comment]
This doesn't sound quite normal. It could be because of insufficient data. I'm not sure if using pre-trained model weights on Moving MNIST as initialization would be helpful... Your data seems to be similar to Moving MNIST...

**Comment by sheilabam** [comment]
Thanks for your help, I'll actively try it.

**Comment by chengtan9907** [comment]
I will close this issue if there is no more question. Please feel free to open a new issue when you have new questions.
